---
title: "Testing Environment"
author: "Caroline Sullivan"
date: "2023-05-03"
output: html_document
---

```{r setup, include=FALSE}
source("analysis.R")
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{r getting metadata}
meta <- get_md()
meta_filtered <- filter_md(meta)
head(meta_filtered)
```
```{r summarizing metadata}
meta_summary <- summarize_md(meta_filtered)
meta_summary
```
```{r visualizing metadata}
barplot <- visualize_md(meta_filtered, 'Lifestage', 'Timepoint')
barplot
```

```{r prep counts data}
counts <- get_data(meta)

male_counts = counts[1][[1]]
female_counts = counts[2][[1]]
larvae_counts = counts[3][[1]]
all_counts = counts[4][[1]]

head(counts['All'])

```
```{r test filters}
counts_filtered <- filter_counts(all_counts, 0.5, 20)

counts_filtered

```
```{r scatter1}
scatter1 <- scatter_plot1(all_counts, 0.5, 20)

scatter1

```
```{r scatter2}
scatter2 <- scatter_plot2(all_counts, 0.5, 20)

scatter2

```

```{r heatmap}
counts_heatmap <- plot_heatmap(all_counts, meta, 0.9, 20)
print(counts_heatmap)

```

```{r pca}
pca_biplot <- plot_pca(all_counts, meta, 1, 2)
print(pca_biplot)

```

```{r DESeq2}
deseq_results_male <- run_deseq(male_counts, get_col_data(male_counts, meta, 'Male'), 
                                10, "condition_fluctuating_vs_control")

deseq_results_female <- run_deseq(female_counts, get_col_data(female_counts, meta, 'Female'),
                                  10, "condition_fluctuating_vs_control")

deseq_results_larvae <- run_deseq(larvae_counts, get_col_data(larvae_counts, meta, 'Larvae'), 
                                 10, "condition_fluctuating_vs_control")

```

```{r load DESeq2 res}
deseq_results_male <- readRDS('objects/deseq_m_fluctvsctrl')

deseq_results_female <- readRDS('objects/deseq_f_fluctvsctrl')

deseq_results_larvae <- readRDS('objects/deseq_l_fluctvsctrl')

```

```{r DESeq2 results}
head(deseq_results_larvae)
```

```{r DESeq2 comparison}

gene_names <- list(Male = de_genes(deseq_results_male, -2), 
                   Female = de_genes(deseq_results_female, -2), 
                   Larvae = de_genes(deseq_results_larvae, -2))

ggVennDiagram(x=gene_names) +
  ggtitle('Differentially Expressed Transcripts fluctuating vs. control') +
   scale_x_continuous(expand = expansion(mult = .2))

```

```{r volcano plot}
volc_plot <- volcano(as.data.frame(readRDS('objects/deseq_f_fluctvsctrl')), 
                     'log2FoldChange',
                     'padj',
                     -10,
                     'black',
                     'lightblue')
                     
print(volc_plot)

```

```{r clustering method 1}
clustering_female <- clustering('Female', female_counts, meta, deseq_results_female, -5, 40)
                     
```
``` {r cluster plot}

cluster_full_plot(clustering_female, "Female")

```
```{r timecourse heatmap}

female_timecourse <- timecourse_heatmap(clustering_female, female_counts, "Female")
   
```
```{r plotting counts data for indivdual genes}

cluster_gene_plot2(clustering_female, 
                   "Female", 
                   get_col_data(female_counts, meta, "Female"),
                   'FBgn0003068')

```

```{r clustering method 2}
clustering_malev2 <- clustering_masigpro(male_counts, 
                                         get_col_data(male_counts, meta, "Male"), 
                                         2, 9)
                     

```

